lib/ansible/utils/vars.py

```diff
@@ -92,32 +92,90 @@ def combine_vars(a, b):
         return result
 
 
-def merge_hash(a, b):
+def merge_hash(x, y, recursive=True, list_merge='keep'):
     """
-    Recursively merges hash b into a so that keys from b take precedence over keys from a
+    Return a new dictionary result of the merges of y into x,
+    so that keys from y take precedence over keys from x.
+    (x and y aren't modified)
     """
-
-    _validate_mutable_mappings(a, b)
-
-    # if a is empty or equal to b, return b
-    if a == {} or a == b:
-        return b.copy()
-
-    # if b is empty the below unfolds quickly
-    result = a.copy()
-
-    # next, iterate over b keys and values
-    for k, v in iteritems(b):
-        # if there's already such key in a
-        # and that key contains a MutableMapping
-        if k in result and isinstance(result[k], MutableMapping) and isinstance(v, MutableMapping):
-            # merge those dicts recursively
-            result[k] = merge_hash(result[k], v)
-        else:
-            # otherwise, just copy the value from b to a
-            result[k] = v
-
-    return result
+    if list_merge not in ('keep', 'append', 'prepend', 'append_np', 'prepend_np'):
```

Comment:
might want a `replace` similar to `keep`, and bikeshedding on the names

Reply:
@jamescassell done